qstarz_bl_1000.cc

```diff
@@ -0,0 +1,315 @@
+    }
+
+    break;
+  }
+
+  auto* waypoint = new Waypoint();
+
+  int tmp_lat = dLatitude / 100;
+  int tmp_lon = dLongitude / 100;
+
+  waypoint->latitude = tmp_lat + (dLatitude - tmp_lat * 100) / 60.0;
+  waypoint->longitude = tmp_lon + (dLongitude - tmp_lon * 100) / 60.0;
+
+  // printf(MYNAME ": waypoint->latitude: %f\n", waypoint->latitude);
+  // printf(MYNAME ": waypoint->longitude: %f\n", waypoint->longitude);
+
+  if ((waypoint->latitude < -90) || (waypoint->latitude > 90)) {
+    fatal(MYNAME ": File format error on '%s'. Unexpected value for latitude: %f\n", qPrintable(read_fname), waypoint->latitude);
+  }
+
+  if ((waypoint->longitude < -180) || (waypoint->longitude > 180)) {
+    fatal(MYNAME ": File format error on '%s'. Unexpected value for longitude: %f\n", qPrintable(read_fname), waypoint->longitude);
+  }
+
+  waypoint->altitude = altitude;
+
+  waypoint->hdop = round(hdop * 1000) / 1000;
+  waypoint->vdop = round(vdop * 1000) / 1000;
+  waypoint->fix = fix;
+  waypoint->sat = satelliteCountUsed;
+
+  waypoint->speed = speed;
+  waypoint->wpt_flags.speed = 1;
+
+  waypoint->course = heading;
+  waypoint->wpt_flags.course = 1;
+
+  double timeIntervalSince1970 = time + milliseconds / 1000.0;
+  QDateTime dateTime = QDateTime::fromSecsSinceEpoch(timeIntervalSince1970);
+
+  waypoint->SetCreationTime(gpsbabel::DateTime(dateTime));
+
+  qstarz_bl_1000_fsdata* fsdata = qstarz_bl_1000_alloc_fsdata();
+
+  fs_chain_add(&(waypoint->fs), (format_specific_data*) fsdata);
+
+  fsdata->accelerationOffset = 0;
+  fsdata->accelerationX = gx / 256.0;
+  fsdata->accelerationY = gy / 256.0;
+  fsdata->accelerationZ = gz / 256.0;
+
+  fsdata->satTotal = satelliteCountView;
+  fsdata->maxSNR = maxSNR;
+  fsdata->batteryPercent = batteryPercent;
+
+  if (type == BL1000_POINT_TYPE_WAY_POINT) {
+    if (global_opts.synthesize_shortnames) {
+      waypoint->shortname = QString("WP%2").arg(waypt_count() + 1, 3, 10, QChar('0'));
+      waypoint->wpt_flags.shortname_is_synthetic = 1;
+    }
+    waypt_add(waypoint);
+  } else if ((type == BL1000_POINT_TYPE_TRACK_POINT_BY_TIME) || (type == BL1000_POINT_TYPE_TRACK_POINT_BY_DISTANCE) || (type == BL1000_POINT_TYPE_TRACK_POINT_BY_SPEED)) {
+    track_add_wpt(track_route, waypoint, "TP", 3);
+  } else {
+    delete waypoint;
+  }
+}
+
+/* fsdata manipulation functions */
+void
+qstarz_bl_1000_free_fsdata(void* fsdata)
+{
+  xfree(fsdata);
+}
+
+void
+qstarz_bl_1000_copy_fsdata(qstarz_bl_1000_fsdata** dest, qstarz_bl_1000_fsdata* src)
+{
+  *dest = (qstarz_bl_1000_fsdata*)xmalloc(sizeof(*src));
+  ** dest = *src;
+  (*dest)->fs.next = nullptr;
+}
+
+qstarz_bl_1000_fsdata*
+qstarz_bl_1000_alloc_fsdata()
+{
+  auto* fsdata = (qstarz_bl_1000_fsdata*) xcalloc(1, sizeof(qstarz_bl_1000_fsdata));
+  fsdata->fs.copy = (fs_copy) qstarz_bl_1000_copy_fsdata;
+  fsdata->fs.destroy = qstarz_bl_1000_free_fsdata;
+
+  return fsdata;
+}
```

Comment:
This needs to catch up with base class FormatSpecificData.  qstarz_bl_1000_fsdata should be a derived class of this.  Use new instead of xmalloc/xcalloc to line up with FormatSpecificDataList::FsChainDestroy().

Reply:
Where can I find information on FormatSpecificData? Specifically, it seems I need to define an FSTYPE